### PR TITLE
Functionnal test

### PR DESCRIPTION
**Pull request by AdrienMQ**
These patches add the functionnal tests to the parameter-framework.
They also dissociate the make install and the make test.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -38,17 +38,22 @@ project(parameter-framework)
...
+
 option(COVERAGE "Build with coverage support" OFF)
 if(COVERAGE)
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fprofile-arcs -ftest-coverage")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -g -O0 -fprofile-arcs -ftest-coverage")
     set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -fprofile-arcs -ftest-coverage")
 endif()
 
+set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
+set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib)
```
Comment (dawagner) [inline CMakeLists.txt]: duplicate

File: bindings/c/CMakeLists.txt [hunk 2]
```diff
@@ -56,5 +56,8 @@ if(BUILD_TESTING)
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-nonnull")
 
     target_link_libraries(cparameterUnitTest cparameter)
-    add_test(cparameterUnitTest cparameterUnitTest)
+    add_test(NAME cparameterUnitTest
+             COMMAND sh -c "PATH=${CMAKE_RUNTIME_OUTPUT_DIRECTORY}:$PATH \\
+                            LD_LIBRARY_PATH=${CMAKE_LIBRARY_OUTPUT_DIRECTORY}:$LD_LIBRARY_PATH \\
+                            cparameterUnitTest")
```
Comment (dawagner) [inline bindings/c/CMakeLists.txt]: use set_property for setting the test's environment

File: test/test-platform/CMakeLists.txt [hunk 3]
```diff
@@ -26,18 +26,20 @@
...
-
-include_directories(
-    "${PROJECT_SOURCE_DIR}/parameter/include"
-    "${PROJECT_SOURCE_DIR}/remote-processor"
-    "${PROJECT_SOURCE_DIR}/utility")
-
-target_link_libraries(test-platform parameter remote-processor)
-
-install(TARGETS test-platform RUNTIME DESTINATION bin)
+if (BUILD_TESTING)
```
Comment (dawagner) [inline test/test-platform/CMakeLists.txt]: don't condition the build of test-platform: it can be used for something else than testing the parameter-framework

File: test/test-subsystem/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    include_directories(
+        include
+        "${PROJECT_SOURCE_DIR}/utility"
+        "${PROJECT_SOURCE_DIR}/xmlserializer"
+        "${PROJECT_SOURCE_DIR}/remote-processor"
+        "${PROJECT_SOURCE_DIR}/remote-process"
+        "${PROJECT_SOURCE_DIR}/parameter"
+        "${PROJECT_SOURCE_DIR}/parameter/include") 
+    
+    target_link_libraries(test-subsystem parameter remote-processor)
```
Comment (dawagner) [inline test/test-subsystem/CMakeLists.txt]: I don't think libtest-subsystem needs to link against libremote-processor.

File: test/test-subsystem/TESTSubsystem.cpp [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+    // Provide creators to upper layer
+    addSubsystemObjectFactory(new TSubsystemObjectFactory<CTESTSubsystemBinary>("Binary", 1 << ETESTDirectory));
+    addSubsystemObjectFactory(new TSubsystemObjectFactory<CTESTSubsystemString>("String", 1 << ETESTDirectory));
+}
+
+// Susbsystem sanity health
+bool CTESTSubsystem::isAlive() const
+{
+    return read(std::string(gacFwNamePropName) + "/isAlive") == "true";
```
Comment (dawagner) [inline test/test-subsystem/TESTSubsystem.cpp]: `gacFwNamePropName` could be `NULL` which would cause an unmanaged exception. Please assert that `gacFwNamePropName != NULL`

File: test/test-subsystem/TESTSubsystem.cpp [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+// Resynchronization after subsystem restart needed
+bool CTESTSubsystem::needResync(bool bClear)
+{
+    std::string strNeedResyncFile = std::string(gacFwNamePropName) + "/needResync";
```
Comment (dawagner) [inline test/test-subsystem/TESTSubsystem.cpp]: ditto

File: test/test-subsystem/TESTSubsystem.h [hunk 7]
```diff
@@ -0,0 +1,50 @@
+/*
+* Copyright (c) 2011-2014, Intel Corporation
```
Comment (dawagner) [inline test/test-subsystem/TESTSubsystem.h]: 2011-2015

File: test/test-subsystem/TESTSubsystemBinary.cpp [hunk 8]
```diff
@@ -0,0 +1,64 @@
+/*
+* Copyright (c) 2011-2014, Intel Corporation
```
Comment (dawagner) [inline test/test-subsystem/TESTSubsystemBinary.cpp]: ditto

File: test/test-subsystem/TESTSubsystemBinary.h [hunk 9]
```diff
@@ -0,0 +1,47 @@
+/*
+* Copyright (c) 2011-2014, Intel Corporation
```
Comment (dawagner) [inline test/test-subsystem/TESTSubsystemBinary.h]: ditto

File: test/test-subsystem/TESTSubsystemBuilder.cpp [hunk 10]
```diff
@@ -0,0 +1,42 @@
+/*
+* Copyright (c) 2011-2014, Intel Corporation
```
Comment (dawagner) [inline test/test-subsystem/TESTSubsystemBuilder.cpp]: ditto

File: test/test-subsystem/TESTSubsystemObject.cpp [hunk 11]
```diff
@@ -0,0 +1,150 @@
+/*
+* Copyright (c) 2011-2014, Intel Corporation
```
Comment (dawagner) [inline test/test-subsystem/TESTSubsystemObject.cpp]: ditto

File: test/test-subsystem/TESTSubsystemObject.h [hunk 12]
```diff
@@ -0,0 +1,59 @@
+/*
+* Copyright (c) 2011-2014, Intel Corporation
```
Comment (dawagner) [inline test/test-subsystem/TESTSubsystemObject.h]: ditto

File: test/test-subsystem/TESTSubsystemString.cpp [hunk 13]
```diff
@@ -0,0 +1,50 @@
+/*
+* Copyright (c) 2011-2014, Intel Corporation
```
Comment (dawagner) [inline test/test-subsystem/TESTSubsystemString.cpp]: ditto

File: test/test-subsystem/TESTSubsystemString.h [hunk 14]
```diff
@@ -0,0 +1,45 @@
+/*
+* Copyright (c) 2011-2014, Intel Corporation
```
Comment (dawagner) [inline test/test-subsystem/TESTSubsystemString.h]: ditto

File: test/functionnal-tests/ACTCampaignEngine.py [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+    os.environ["PFW_TEST_TOOLS"] = "."
+    os.environ["PFW_TEST_CONFIGURATION"] = args.pfw_fs + "xml/configuration/ParameterFrameworkConfiguration.xml"
+    
+    try:
+        res = testsRunner('PfwTestCase')
+    
+    finally:
+        os.environ["PFW_FILESYSTEM"] = ""
+        os.environ["PFW_TEST_TOOLS"] = ""
+        os.environ["PFW_TEST_CONFIGURATION"] = ""
```
Comment (dawagner) [inline test/functionnal-tests/ACTCampaignEngine.py]: After this, the environment will still contains these variables (they will contain an empty string but still exist nonetheless). Do this instead: `del os.environ["PFW_FILESYSTEM"]`

File: test/test-subsystem/CMakeLists.txt (hunk 4 above)
Comment (AdrienMQ) [inline test/test-subsystem/CMakeLists.txt]: conserve only the parameter and remove remove-process

File: test/functionnal-tests/ACTCampaignEngine.py (hunk 15 above)
Comment (AdrienMQ) [inline test/functionnal-tests/ACTCampaignEngine.py]: Remove try/finally.
Env vars are deleted when the script end.

File: CMakeLists.txt (hunk 1 above)
Comment (AdrienMQ) [inline CMakeLists.txt]: done

File: test/test-platform/CMakeLists.txt (hunk 3 above)
Comment (AdrienMQ) [inline test/test-platform/CMakeLists.txt]: done

File: test/test-subsystem/TESTSubsystem.cpp (hunk 5 above)
Comment (AdrienMQ) [inline test/test-subsystem/TESTSubsystem.cpp]: done

File: test/test-subsystem/TESTSubsystem.cpp (hunk 6 above)
Comment (AdrienMQ) [inline test/test-subsystem/TESTSubsystem.cpp]: done

File: bindings/c/CMakeLists.txt (hunk 2 above)
Comment (AdrienMQ) [inline bindings/c/CMakeLists.txt]: done

**Comment by dawagner** [comment]
some copyright years still need to be updated

**Comment by dawagner** [comment]
Please clean all whitespaces by running `git rebase -f --whitespace=fix v2.4.2`

File: bindings/c/CMakeLists.txt [hunk 16]
```diff
@@ -56,5 +56,11 @@ if(BUILD_TESTING)
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-nonnull")
 
     target_link_libraries(cparameterUnitTest cparameter)
-    add_test(cparameterUnitTest cparameterUnitTest)
+    add_test(NAME cparameterUnitTest
+             COMMAND cparameterUnitTest)
```
Comment (dawagner) [inline bindings/c/CMakeLists.txt]: is this change really needed ?

File: test/functionnal-tests/ACTCampaignEngine.py [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+
+These commands are tested using the methods of the classes
+"BooleanTestCase", etc...
+"""
+
+import os, sys
+import unittest
+import argparse
+import shutil
+import tempfile
```
Comment (dawagner) [inline test/functionnal-tests/ACTCampaignEngine.py]: argparse, shutil and tempfile aren't used.

File: test/functionnal-tests/ACTCampaignEngine.py [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+def testsRunner(testDirectory):
+
+    tests = unittest.defaultTestLoader.discover(testDirectory, pattern='t*.py')
+    runner = unittest.TextTestRunner(verbosity=2)
+
+    return runner.run(tests).wasSuccessful()
+
+def main():
+
+    os.environ["PFW_FILESYSTEM"] = os.environ["PFW_FS"]
```
Comment (dawagner) [inline test/functionnal-tests/ACTCampaignEngine.py]: I'm worried that 2 consecutive runs of the functional tests will reuse the same files and the second run might be poisoned. I'm not sure this is a problem.

File: test/functionnal-tests/CMakeLists.txt [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+        ${PFW_FS}xml/configuration/ParameterFrameworkConfiguration.xml
+        COPYONLY)
+    configure_file(
+        ${CMAKE_CURRENT_SOURCE_DIR}/xml/configuration/Settings/Test/TestConfigurableDomains.xml
+        ${PFW_FS}xml/configuration/Settings/Test/TestConfigurableDomains.xml
+        COPYONLY)
+
+    add_test(NAME functionnal-test 
+             WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR} 
+             COMMAND ${python2} ACTCampaignEngine.py)
```
Comment (dawagner) [inline test/functionnal-tests/CMakeLists.txt]: When this command is run, a lot of ".pyc" files are generated in the sources. It seems important to me that this is avoided.

**Comment by dawagner** [comment]
See a77c5af82c621b7082dd22c60af137c84a14b5a1 for how to avoid valgrind/memcheck to be run on the functional tests; which takes a lot of time and is useless.

File: test/functionnal-tests/README.md [hunk 20]
```diff
@@ -0,0 +1,15 @@
+#FUNCTIONNAL TEST
```
Comment (dawagner) [inline test/functionnal-tests/README.md]: no need for ALL CAPS

File: test/functionnal-tests/README.md [hunk 21]
```diff
@@ -0,0 +1,15 @@
...
+
+##What ?
+Functionnal tests are unit tests that ensure the functionment of the parameter-framework.
```
Comment (dawagner) [inline test/functionnal-tests/README.md]: be careful: functionnal test are *not* unit tests. They are conceptually different

File: test/functionnal-tests/README.md [hunk 22]
```diff
@@ -0,0 +1,15 @@
...
+Types,functions and Domains are tested.
+
+##Practical
+
+Tests can be lunch with the 'make test' command.
```
Comment (dawagner) [inline test/functionnal-tests/README.md]: launched

File: test/functionnal-tests/README.md [hunk 23]
```diff
@@ -0,0 +1,15 @@
...
+The test flag in cmake is set to 'ON' by default.
```
Comment (dawagner) [inline test/functionnal-tests/README.md]: not relevant.

File: test/functionnal-tests/README.md [hunk 24]
```diff
@@ -0,0 +1,15 @@
...
+
+Environment vars are set in the cmake.
+Then we lunch ACTCampaignEngine.py which execute all the tests one by one.
```
Comment (dawagner) [inline test/functionnal-tests/README.md]: launch

File: test/test-subsystem/CMakeLists.txt [hunk 25]
```diff
@@ -0,0 +1,51 @@
...
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+# ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+if (BUILD_TESTING)
+
+    set(PFW_FS "${CMAKE_BINARY_DIR}/tmp/test-parameters/")
+    file(WRITE "${PFW_FS}/isAlive" "true")
+    file(WRITE "${PFW_FS}/needResync" "false")
```
Comment (dawagner) [inline test/test-subsystem/CMakeLists.txt]: I believe these 3 commands should be part of the functional test's makefile; not here.

**Comment by dawagner** [comment]
`s/functionnal/functional/` (everywhere)

File: test/functionnal-tests/ACTCampaignEngine.py (hunk 18 above)
Comment (AdrienMQ) [inline test/functionnal-tests/ACTCampaignEngine.py]: Now, we remove the test-parameter folder at the end of the tests (we use shutils.rmtree)

File: test/functionnal-tests/CMakeLists.txt (hunk 19 above)
Comment (AdrienMQ) [inline test/functionnal-tests/CMakeLists.txt]: Done
But in counterparty, it seems that these files represent a time gain while executing tests. .pyc file are loaded faster than .py files.

File: test/test-subsystem/CMakeLists.txt (hunk 25 above)
Comment (AdrienMQ) [inline test/test-subsystem/CMakeLists.txt]: done

File: test/functional-tests/ACTCampaignEngine.py [hunk 26]
```diff
@@ -0,0 +1,94 @@
...
+            dirs.remove('xml')
+        for name in files:
+            os.remove(os.path.join(path, name))
+        for name in dirs:
+            os.rmdir(os.path.join(path, name))
+        
+    sys.exit(not res)
+
+if __name__ == "__main__":
+    main()
```
Comment (dawagner) [inline test/functional-tests/ACTCampaignEngine.py]: 1. When creating file paths, use os.path.join instead of "+" betwteen strings, which may be wrong (e.g. "/some/path" + "someFileName" == /some/pathsomeFileName" whereas os.path.join("/some/path", "someFileName") == "/some/path/someFileName")
2. remove the commented line
3. remove surrounding spaces around ˋos.environ["PFW_FS"]ˋ

File: test/functional-tests/CMakeLists.txt [hunk 27]
```diff
@@ -0,0 +1,58 @@
...
+    add_test(NAME functional-test
+             WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+             COMMAND ${python2} -O ACTCampaignEngine.py)
+
+    set_property(TEST functional-test PROPERTY ENVIRONMENT
+            PATH=${CMAKE_RUNTIME_OUTPUT_DIRECTORY}:$ENV{PATH}
+            LD_LIBRARY_PATH=${CMAKE_LIBRARY_OUTPUT_DIRECTORY}:$ENV{LD_LIBRARY_PATH}
+            PFW_FS=${PFW_FS})
+
+endif()
```
Comment (dawagner) [inline test/functional-tests/CMakeLists.txt]: you could set the PFW_FILESYSTEM env var directly

**Comment by dawagner** [comment]
did you run the "git rebase -f --whitespace=fix" command ? (I can't check by myself right now).

**Comment by dawagner** [comment]
The "for" loop in ACTCampaignEngine has several issues:

- if any exception is raised in testRunner, it will not be executed (need for a try...finally block)
- it isn't documented
- I'm not sure why you don't remove the whole PFW_FS directory.
-

**Comment by AdrienMQ** [comment]
I just seen your comments. 
So, I think I have taken all your comments into account.

=> Removed some includes dir in test-subsystem
=> Run "git rebase -f --whitespace=fix"
=> For the copyright, it's done but we need to update all of them (including in python files) within a new patch.
=> I didn't remove the wall pfw path because of the xml temporary file which is needed to relaunch the test. I have implemented the try finally for the removal.


I have also changed the name of the environment variables. PFW_FS become PFW_ROOT and PFW_FILESYSTEM become PFW_RESULT (PFW_RESULT=PFW_ROOT/result).
Now, we have the xml directory in PFW_ROOT and I redirect the output on PFW_RESULT.
I remove PFW_RESULT at the end of the test but we conserve the xml files to have the availability to relaunch the tests.

Now, I have a persistant error which is : ParameterMgr: Unable to start remote processor server on port 5000. It seems to be a random error. I'm working on this, I'll push my changes when this will be fixed.

Thank you for your review. :+1:

**Comment by AdrienMQ** [comment]
:+1:

**Comment by dawagner** [comment]
I see the following log in several places in the functional test logs:

```
Please provide absolute path
```

This seems like an error but the test cases pass nonetheless; there must be an issue with these tests cases (probaly tied to the fact that `PFW_TEST_TOOLS = "."`.

Another error:

```
3: test_Restore_Config_Error (Domains.tDomain_Configuration_Backup.TestCases) ... Command not found!
```

here, the test is supposed to be checking that an error is returned when there is a missing argument to the command, *not if the command name is wrong* (indeed, the `RestoreConfiguration` command is called instead of `restoreConfiguration`

Please check all errors; especially the test cases that used to fail.

I've also left some comments on f851cb63db821bd9d9e5a403bfc7904cc207ae85

You're almost there :)

**Comment by AdrienMQ** [comment]
I have fixed your comments bit it seems to be some errors related in the log : tFunctions_ListingFunction, tBit_Block ....

**Comment by AdrienMQ** [comment]
Update last modifications.
According to J. Clero, some test are not fixed and must fail. With the changes brought by this pull request, it is hard to verify manually.

**Comment by AdrienMQ** [comment]
Not a diff between local and remote version but a bug in the visualisation?
